File: vault/src/password/hasher.java
package password;

public class hasher {
    private final int size;

    public hasher(int size) {
        this.size = size;
    }

    public int gen_pass(String password) {
        int hash = 0;
        final int PRIME = 31;

        if (password == null) {
            return 0;
        }

        for (int i = 0; i < password.length(); i++) {
            char ch = password.charAt(i);
            hash = PRIME * hash + ch;
        }
        int index = hash % size;

        if (index < 0) {
            index += size;
        }
        return index;
    }

    public int gen_user(String username) {
        int hash = 0;
        final int PRIME = 31;

        for (int i = 0; i < username.length(); i++) {
            hash = hash * PRIME + username.charAt(i);
        }
        return Math.abs(hash);
    }
}

File: vault/algorithms/hasher/hasher.cpp
//
// Created by nicol on 8/20/2025.
//

#include "hasher.h"

int hashing::multiplication(const std::string& key, int table) {
    long long hash = 0;

    for (char ch : key) {
        constexpr int PRIME = 31;

        hash = hash * PRIME + ch;
    }

    int index = hash % table;

    if (index < 0) index += table;

    return index;
}

int hashing::mid_square(const std::string& key) {
    long long hash = 0;

    for (char ch : key) {
        constexpr int PRIME = 31;

        hash = hash * PRIME + ch;
    }
    return std::abs(static_cast<int>(hash % 2147483647));
}